examples/allocation/zig/greet.go

```diff
@@ -0,0 +1,142 @@
+	if err != nil {
+		return err
+	}
+
+	// Note: testdata/greet.zig doesn't use WASI, but Zig needs it to
+	// implement functions such as panic.
+	if _, err = wasi_snapshot_preview1.Instantiate(ctx, r); err != nil {
+		return err
+	}
+
+	// Instantiate a WebAssembly module that imports the "log" function defined
+	// in "env" and exports "memory" and functions we'll use in this example.
+	compiled, err := r.CompileModule(ctx, greetWasm, wazero.NewCompileConfig())
+	if err != nil {
+		return err
+	}
+
+	mod, err := r.InstantiateModule(ctx, compiled, wazero.NewModuleConfig().WithStdout(os.Stdout).WithStderr(os.Stderr))
+	if err != nil && isFailureExit(err) {
+		return err
+	}
+
+	// Get references to WebAssembly functions we'll use in this example.
+	greet := mod.ExportedFunction("greet")
+	greeting := mod.ExportedFunction("greeting")
+
+	malloc := mod.ExportedFunction("malloc")
+	free := mod.ExportedFunction("free")
+
+	// Let's use the argument to this main function in Wasm.
+	name := os.Args[1]
+	nameSize := uint64(len(name))
+
+	// Instead of an arbitrary memory offset, use Zig's allocator. Notice
+	// there is nothing string-specific in this allocation function. The same
+	// function could be used to pass binary serialized data to Wasm.
+	results, err := malloc.Call(ctx, nameSize)
+	if err != nil && isFailureExit(err) {
+		return err
+	}
+	namePtr := results[0]
+	if namePtr == 0 {
+		return errors.New("malloc failed")
+	}
+	// We have to free this pointer when finished.
+	defer free.Call(ctx, namePtr, nameSize)
+
+	// The pointer is a linear memory offset, which is where we write the name.
+	if !mod.Memory().Write(ctx, uint32(namePtr), []byte(name)) {
+		return fmt.Errorf("Memory.Write(%d, %d) out of range of memory size %d",
+			namePtr, nameSize, mod.Memory().Size(ctx))
+	}
+
+	// Now, we can call "greet", which reads the string we wrote to memory!
+	_, err = greet.Call(ctx, namePtr, nameSize)
+	if err != nil && isFailureExit(err) {
+		return err
+	}
+
+	// Finally, we get the greeting message "greet" printed. This shows how to
+	// read-back something allocated by Zig.
+	ptrSize, err := greeting.Call(ctx, namePtr, nameSize)
+	if err != nil && isFailureExit(err) {
+		return err
+	}
+
+	greetingPtr := uint32(ptrSize[0] >> 32)
+	greetingSize := uint32(ptrSize[0])
+	// The pointer is a linear memory offset, which is where we write the name.
+	if bytes, ok := mod.Memory().Read(ctx, greetingPtr, greetingSize); !ok {
+		return fmt.Errorf("Memory.Read(%d, %d) out of range of memory size %d",
+			greetingPtr, greetingSize, mod.Memory().Size(ctx))
+	} else {
+		fmt.Println("go >>", string(bytes))
+	}
+
+	return nil
+}
+
+// Looks like Zig exits the module after calling each exported function
+// regardless of whether there was an error. We need to handle exit code 0
+// so that the subsequent functions can be called.
+func isFailureExit(err error) bool {
+	if exitErr, ok := err.(*sys.ExitError); !ok || exitErr.ExitCode() != 0 {
```

Comment:
hmm this is a problem I think because state in wazero is changed when the module exits as this is an example, I don't think we should encourage people to use a module after it is freed. Also it is quite surprising to have something exit on account of a random exported function use. Ack WASI isn't well specified, but this isn't de'facto behavior either.

I think the ways out are:
* Change this comment to a TODO, which warns the person this is not something they should do and will break
* Change to reactor model and remove this comment
* Change zig to not exit after each exported function call, as I don't think any other lib does this so it can lead to problems.
* Add a hack option in wazero to ignore exit and use that here

wdyt @mathetake @Luukdegram

Reply:
I recall this can be avoided by compiling this as the shared library instead of exe, but my memory on zig is fuzzy now

Reply:
or, as I said and limit this example to wasi, we can use `reactor` https://github.com/mathetake/proxy-wasm-zig-sdk/blob/main/build.zig#L16

Reply:
The reason why the module exits after execution is because the `build.zig` file is creating an executable. We use the 'command' execution model by default in such case. This means there's 3 options that can be used to solve this:

- Use the reactor model instead.
- Build a shared library in the `build.zig` file: `b.addSharedLibrary("name", .unversioned);`
- Overwrite the startup function by exporting your own _start using `pub export fn _start() void{}`.

Of those 3 options, the most correct one would be to build a shared library as those examples don't necessarily use WASI.

Reply:
Thank you, guys. I've switched to lib.
https://github.com/tetratelabs/wazero/pull/741/commits/b9018858415e81bb1f0aa4473748624b5585772d

Reply:
I think it's also possible to use `exe.wasi_exec_model = .reactor;` on build.zig to use reactor and prevent exit. _If I remember correctly, was @mathetake who add that feature._ 😅